File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/ViewModel/ProfileMessageViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase
import Utils

@Factory
public final class ProfileMessageViewModel: StateProducingViewModel<ProfileMessageViewModel.State, MessageError> {
    @DIArgument private let dog: Dog
    private let sendMessageUseCase: SendMessageUseCase

    init(
        dog: Dog,
        sendMessageUseCase: SendMessageUseCase
    ) {
        self.dog = dog
        self.sendMessageUseCase = sendMessageUseCase
        super.init(initialValue: .initial)
    }

    public func onTextChange(_ text: String) {
        setState(currentState.copy(text: text))
    }

    public func onSendTap() {
        let text = currentState.text
        sendMessageUseCase(dog: dog, text: text)
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success: setState(currentState.copy(text: "", toast: .messageSent(name: dog.name)))
                case .failure(let error): emitError(error)
                }
            }
            .store(in: &cancellables)
    }

    public func onToastDismiss() {
        setState(currentState.copy(toast: nil))
    }
}

extension ProfileMessageViewModel {
    public struct State: Equatable {
        public let text: String
        public let toast: ProfileToastUIModel?

        public var isSendEnabled: Bool {
            !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }

        public static let initial = State(text: "", toast: nil)

        public init(text: String, toast: ProfileToastUIModel?) {
            self.text = text
            self.toast = toast
        }

        func copy(text: String? = nil, toast: ProfileToastUIModel?? = nil) -> State {
            State(text: text ?? self.text, toast: toast ?? self.toast)
        }
    }
}

File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfileMessageViewModel.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.presentation.profile.uimodel.ProfileToastUIModel
import com.perrystreet.woof.usecase.messages.SendMessageUseCase
import io.reactivex.rxjava3.kotlin.plusAssign
import org.koin.android.annotation.KoinViewModel
import org.koin.core.annotation.InjectedParam
import java.util.Optional

@KoinViewModel
class ProfileMessageViewModel(
    @InjectedParam private val dog: Dog,
    private val sendMessageUseCase: SendMessageUseCase,
) : StateProducingViewModel<ProfileMessageViewModel.State>(initialValue = State.Initial) {

    data class State(
        val text: String,
        val toast: ProfileToastUIModel?,
    ) {
        val isSendEnabled: Boolean
            get() = text.isNotBlank()

        companion object {
            val Initial = State(text = "", toast = null)
        }
    }

    fun onTextChange(text: String) {
        _state.onNext(currentState.copy(text = text))
    }

    fun onSendTap() {
        val text = currentState.text
        disposables += sendMessageUseCase(dog, text).subscribe(
            { _state.onNext(currentState.copy(text = "", toast = ProfileToastUIModel.MessageSent(dog.name))) },
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }

    fun onToastDismiss() {
        _state.onNext(currentState.copy(toast = null))
    }
}
